modules/core/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3DirectBatchHandler.java

```diff
@@ -258,29 +257,28 @@ public boolean completeUpload(String batchId, String fileIndex, BatchFileInfo fi
         blobInfo.filename = fileInfo.getFilename();
         blobInfo.length = metadata.getContentLength();
 
+        CopyObjectRequest copyObjectRequest = new CopyObjectRequest(bucket, fileKey, bucket, bucketKey);
         // server-side encryption
-        String targetSSEAlgorithm;
         if (useServerSideEncryption) { // TODO KMS
-            targetSSEAlgorithm = ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION;
-        } else {
-            targetSSEAlgorithm = null;
+            // SSE-S3
+            ObjectMetadata newObjectMetadata = new ObjectMetadata();
+            newObjectMetadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+            copyObjectRequest.setNewObjectMetadata(newObjectMetadata);
         }
 
-        ObjectMetadata newMetadata;
-        if (metadata.getContentLength() > lowerThresholdToUseMultipartCopy()) {
-            newMetadata = S3Utils.copyFileMultipart(amazonS3, metadata, bucket, fileKey, bucket, bucketKey,
-                    targetSSEAlgorithm, true);
-        } else {
-            newMetadata = S3Utils.copyFile(amazonS3, metadata, bucket, fileKey, bucket, bucketKey, targetSSEAlgorithm, true);
-            boolean isMultipartUpload = REGEX_MULTIPART_ETAG.matcher(key).find();
-            if (isMultipartUpload) {
-                key = newMetadata.getETag();
-                String previousBucketKey = bucketKey;
-                bucketKey = bucketPrefix + key;
-                newMetadata = S3Utils.copyFile(amazonS3, metadata, bucket, previousBucketKey, bucket, bucketKey, true);
-            }
+        TransferManager transferManager = getTransferManager();
+        Copy copy = transferManager.copy(copyObjectRequest, amazonS3, null);
+        try {
+            copy.waitForCompletion();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new NuxeoException(e);
+        } finally {
+            amazonS3.deleteObject(bucket, fileKey);
         }
 
+        ObjectMetadata newMetadata = amazonS3.getObjectMetadata(bucket, bucketKey);
```

Comment:
Instead of `copy.waitForCompletion` if you call `copy.waitForCopyResult` it will a return a `CopyResult` object with a `getETag` method. It's better than doing an additional round-trip but it doesn't do the same as `newMetadata.getContentMD5()`... as the ETag may not be an MD5.
I see internal AWS SDK code that just checks for the presence of `-` to determine if it's an ETag from multipart upload... I think we could just say that if it's 32 chars en length then it's a proper MD5 and we can use it as the digest. Or maybe keep it for now and refactor later.

Reply:
I plan to do another commit where the usage of a generic etag vs md5 is possible behind a flag, will take that into account